### PR TITLE
use anyhow for `evtx_dump` and thiserror for `evtx` library.

### PR DESCRIPTION
**Pull request by omerbenamram**
`anyhow` and `thiserror` are much nicer to use than `snafu`.
They behave closer to how std's error would naturally be.

- This does make backtraces rely on `std::backtrace` support (which is currenly only on nigthly), but it seems to be converging.

### CODE REVIEW

File: src/bin/evtx_dump.rs [hunk 1]
```diff
@@ -225,45 +204,41 @@ impl EvtxDump {
...
         };
 
         Ok(())
     }
 
-    fn try_to_initialize_logging(&self) {
-        if let Some(level) = self.verbosity_level {
-            match simplelog::WriteLogger::init(
+    fn try_to_initialize_logging(&self) -> Result<()> {
+        Ok(if let Some(level) = self.verbosity_level {
```
Comment (ohadravid) [inline src/bin/evtx_dump.rs]: Wow wired

Maybe use the explicit `Ok(())`?

File: src/binxml/assemble.rs [hunk 2]
```diff
@@ -26,21 +24,23 @@ pub fn parse_tokens<T: BinXmlOutput>(
         match owned_token {
             XmlModel::OpenElement(open_element) => {
                 stack.push(open_element);
-                visitor.visit_open_start_element(stack.last().context(
-                    err::FailedToCreateRecordModel {
-                        message: "Invalid parser state - expected stack to be non-empty",
-                    },
-                )?)?
+                visitor.visit_open_start_element(stack.last().ok_or(
```
Comment (ohadravid) [inline src/binxml/assemble.rs]: `ok_or_else(|| ` seems a bit nicer, no need to create the error every time

File: src/binxml/assemble.rs [hunk 3]
```diff
@@ -26,21 +24,23 @@ pub fn parse_tokens<T: BinXmlOutput>(
...
+                let close_element = stack.pop().ok_or(EvtxError::FailedToCreateRecordModel(
+                    "Invalid parser state - expected stack to be non-empty",
+                ))?;
                 visitor.visit_close_element(&close_element)?
             }
             XmlModel::Value(s) => visitor.visit_characters(&s)?,
             XmlModel::EndOfStream => {}
             XmlModel::StartOfStream => {}
+            XmlModel::PI(pi) => visitor.visit_processing_instruction(&pi)?,
+            XmlModel::EntityRef(entity) => visitor.visit_entity_reference(&entity)?,
```
Comment (ohadravid) [inline src/binxml/assemble.rs]: 🎉

File: src/evtx_chunk.rs [hunk 4]
```diff
@@ -46,20 +46,24 @@ pub struct EvtxChunkData {
...
-    pub fn new(data: Vec<u8>, validate_checksum: bool) -> Result<Self> {
+    pub fn new(data: Vec<u8>, validate_checksum: bool) -> EvtxChunkResult<Self> {
         let mut cursor = Cursor::new(data.as_slice());
         let header = EvtxChunkHeader::from_reader(&mut cursor)?;
 
         let chunk = EvtxChunkData { header, data };
-        if validate_checksum {
-            ensure!(chunk.validate_checksum(), err::InvalidChunkChecksum)
+        if validate_checksum && !chunk.validate_checksum() {
+            // TODO: return checksum here.
```
Comment (ohadravid) [inline src/evtx_chunk.rs]: ?

File: src/json_output.rs [hunk 5]
```diff
@@ -97,24 +96,33 @@ impl JsonOutput {
...
+        name: &str,
+    ) -> SerializationResult<()> {
         trace!("insert_node_without_attributes");
         self.stack.push(name.to_owned());
 
         let container =
             self.get_current_parent()
                 .as_object_mut()
-                .context(err::JsonStructureError {
+                .ok_or(SerializationError::JsonStructureError {
```
Comment (ohadravid) [inline src/json_output.rs]: `ok_or_else` here as well (and in the rest of them)

File: src/json_output.rs [hunk 6]
```diff
@@ -237,36 +247,59 @@ impl BinXmlOutput for JsonOutput {
...
+        Err(SerializationError::Unimplemented {
+            message: format!("`{}`: visit_cdata_section", file!()),
+        })
     }
 
-    fn visit_entity_reference(&mut self) -> Result<()> {
-        unimplemented_fn!("visit_entity_reference")
+    fn visit_entity_reference(&mut self, entity: &BinXmlName) -> Result<(), SerializationError> {
+        // JSON doesn't care about entity references.
+        Ok(())
```
Comment (ohadravid) [inline src/json_output.rs]: You don't need to add the actual value of the reference?

File: src/model/raw.rs [hunk 7]
```diff
@@ -46,8 +47,9 @@ impl BinXMLRawToken {
                 more_attributes_expected: true,
             })),
             0x07 | 0x47 => Some(BinXMLRawToken::CDataSection),
-            0x08 | 0x48 => Some(BinXMLRawToken::EntityReference),
-            0x0a | 0x49 => Some(BinXMLRawToken::ProcessingInstructionTarget),
+            0x08 | 0x48 => Some(BinXMLRawToken::CharReference),
+            0x09 | 0x49 => Some(BinXMLRawToken::EntityReference),
```
Comment (ohadravid) [inline src/model/raw.rs]: LOL

File: src/xml_output.rs [hunk 8]
```diff
@@ -112,23 +115,39 @@ impl<W: Write> BinXmlOutput for XmlOutput<W> {
...
+        Ok(())
     }
 
-    fn visit_processing_instruction_target(&mut self) -> Result<()> {
-        unimplemented_fn!("visit_processing_instruction_target")
+    fn visit_character_reference(
+        &mut self,
+        char_ref: Cow<'_, str>,
+    ) -> Result<(), SerializationError> {
+        unimplemented!()
```
Comment (ohadravid) [inline src/xml_output.rs]: Maybe an error?

File: tests/test_record_samples.rs [hunk 9]
```diff
@@ -169,6 +169,60 @@ fn test_event_xml_sample_with_user_data() {
     );
 }
 
+#[test]
+fn test_event_xml_sample_with_entity_ref() {
```
Comment (ohadravid) [inline tests/test_record_samples.rs]: Also generate and test both samples for JSON

File: src/bin/evtx_dump.rs (hunk 1 above)
Comment (omerbenamram) [reply src/bin/evtx_dump.rs]: 👍

File: src/json_output.rs (hunk 5 above)
Comment (omerbenamram) [reply src/json_output.rs]: 👍

File: src/xml_output.rs (hunk 8 above)
Comment (omerbenamram) [reply src/xml_output.rs]: 👍

File: tests/test_record_samples.rs (hunk 9 above)
Comment (omerbenamram) [reply tests/test_record_samples.rs]: done.

File: src/json_output.rs (hunk 6 above)
Comment (omerbenamram) [reply src/json_output.rs]: fixed

**Pull request merged**